### DISCUSSION TITLE
Add functionality to support screen switching.

### DISCUSSION THREAD

**Discussion by MonteCarlos** [cc65/cc65#2490] [Ideas]
I started an MR with an experimental implementation for the C64. https://github.com/cc65/cc65/pull/2489

Although I already took some decisions in the MR, I wanted to discuss several uncertainties about how to implement this.
Eventually the early implementation has to be revised, therefor.

Here is the list about things I would like to clarify, before investing more time into this:

- Is the function to reset the screen really needed, can it be a special case of the switching function ?
- Should we provide low level functions without much error checking or should we provide error checking as well. 
    - E.g. the sanity checks for the screen target or vic bank compliance of screen and charset. 
- Should we integrate save and restore cursor position or should we provide raw switch function and separate functions to deal with cursor positioning? 
  - In the simple first attempt, there were gaps left, when printing, switching, printing and switching back. The reason was that the cursor position advanced independently of the selected screen memory. 
  - Therefor I tended to integrate the cursor positioning.
- Should we allow only original and alternate screen or should it be possible to switch screen memory to several different locations? If allowing several different screen locations, howto deal with cursor positioning?
  - In my attempt I allow only switching to one alternate screen at one time, but if the initialize function is called, then another alternate screen location is allowed, but the cursor position is lost.
- Should the (alternate) screen be cleared after switching? 
  - In my attempt I implemented the screen clearing, but it could be made conditional. 
- Should we add a function to switch the viewport, not only the screen where the writes go? If so, it would also necessarily include the possibility to switch the charset or the bitmap stretching the scope of this mr. This also would imply the need to check VIC bank boundaries and code to switch the VIC bank.
  - On my opinion this function must be added, else we could not display the alternate screen.
- Should we allow positioning the screen at $1000, $9000, $d000, $0000? 
  - In the first two cases, the viewport cannot be switched there. 
  - However, printing there and then doing some memcpy to another location could make sense for e.g. a windowed print. 
  - In the third case, it's possible to position and view a screen there but the typical cc65 memory layout ($36 in $01) is not intended for this. The latter case could have some rare usescases like printing to $200 or $340-$3ff, but it would be very error prone.

**Comment by mrdudz** [comment]
See the PR (i find it super annoying to have comments on a pr in separate discussion for that matter)

**Reply by MonteCarlos** [reply]
Sorry, I though it were a good idea.
Going to move the discussion back into the MR. ;)
